File: find_adjacent.py
import numpy as np
import matplotlib.pyplot as plt
import networkx as nx
import sys

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Please specify xyz input file and adjacency outfile.")
    else:
        inname = sys.argv[1]
        outname = sys.argv[2]


# parses the coordinates from an xyz file
def parse_xyz(xyz_name):
    # make a new datatype (numpy.void)
    xyz_row_format = np.dtype(dtype=[('element', 'U1'), ('x', 'f8'), ('y', 'f8'),('z', 'f8')])
    xyz = np.genfromtxt(xyz_name, skip_header=2, dtype=xyz_row_format)
    return xyz

# throws out all atoms that are not carbon, and returns a numpy array
def keep_carbons(xyz):
    n_cs = 0
    for row in xyz:
        if row[0] == 'C':
            n_cs += 1

    current = 0
    c_xyz = np.empty((n_cs, 3))
    for row in xyz:
        if row[0] == 'C':
            coords = np.array([row[1], row[2], row[3]])
            c_xyz[current] = coords
            current += 1
    
    return c_xyz

# finds distance between two vectors
def find_dist(xyz1, xyz2):
    difference = xyz1 - xyz2
    distance = np.linalg.norm(difference)
    return distance

# builds adjacency matrix, with no cutoffs
def find_adjacency(xyz):
    n_cs = len(xyz)
    adjacency = np.zeros((n_cs, n_cs))
    for i, row1 in enumerate(xyz):
        for j, row2 in enumerate(xyz):
            if i == j:
                continue
            dist = find_dist(row1, row2)
            adjacency[i,j] = dist
    #print(adjacency)
    return adjacency

# build a stripped down connectivity matrix, based on some cutoff
def strip_down(adjacency, cutoff = 2.0):
    adjacency = np.where(adjacency <= cutoff, adjacency, 0)
    adjacency = np.where(adjacency < 0.1, adjacency, 1)
    return adjacency

# plots the adjacency matrix, with some cutoff
def plot_graph(adjacency, cutoff = 2.0):
    graph = nx.from_numpy_matrix(adjacency, create_using=nx.MultiGraph)
    nx.draw(graph, with_labels=True, font_weight='bold')
    plt.savefig("graph.png")
    #plt.show()

# outputs the bonds csv
def output_csv(mat, outname):
    np.savetxt(outname, mat, delimiter=",", fmt='%u')


if __name__ == "__main__":
    xyz = parse_xyz(inname)
    c_xyz = keep_carbons(xyz)
    adjacency = find_adjacency(c_xyz)
    print(adjacency)
    adjacency = strip_down(adjacency)
    plot_graph(adjacency) 
    output_csv(adjacency, outname)

File: find_adjacent.cpp
#include <iostream>
#include <fstream> 
#include <vector>
#include <Eigen/Dense>

void parse_xyz(std::string &, std::vector<std::vector<double>> &, std::vector<std::string> &);
void keep_carbons(std::vector<std::vector<double>> &, std::vector<std::string> &);
double find_dist(std::vector<double> &, std::vector<double> &);
void find_adjacency(std::vector<std::vector<double>> &, Eigen::MatrixXd &);
void strip_down(Eigen::MatrixXd &, double = 2.0);
void output_csv(std::string &, Eigen::MatrixXd &);


int main(int argc, char *argv[]) {
    std::string infile, outfile;
    try {
        if (argc != 3)
            throw std::exception();
        infile = argv[1];
        outfile = argv[2];
    }
    catch(...) {
        std::cout << "Please specify input xyz and output csv for adjacency" << std::endl; 
        return 0;
    }

    std::vector<std::vector<double>> xyz;
    std::vector<std::string> atom_type;
    parse_xyz(infile, xyz, atom_type);
    keep_carbons(xyz, atom_type);
    int n_cs = xyz.size();    
    
    Eigen::MatrixXd adjacency(n_cs, n_cs);
    find_adjacency(xyz, adjacency);
    std::cout << adjacency << std::endl;
    strip_down(adjacency);
    std::cout << adjacency << std::endl;
    output_csv(outfile, adjacency);
}


void parse_xyz(std::string &infile, std::vector<std::vector<double>> &xyz, std::vector<std::string> &atom_type) {
    std::ifstream xyzstream (infile, std::ifstream::in);
    
    std::string line;
    //std::getline(xyz,line, '\r'); // first line
    //int natoms = std::stod(line); // contains number of atoms
    int natoms;
    xyzstream >> natoms;
    std::getline(xyzstream,line, '\r'); // removes second line (comment line)
    std::getline(xyzstream,line, '\r'); // removes second line (comment line)
    std::vector<double> coords(3);

    atom_type.resize(natoms);
    xyz.resize(natoms);

    for (size_t i = 0; i < natoms; i++){
        xyzstream >> atom_type[i] >> coords[0] >> coords[1] >> coords[2];
        xyz[i] = coords;
    }
}

void keep_carbons(std::vector<std::vector<double>> &xyz, std::vector<std::string> &atom_type) {
    size_t natoms = atom_type.size();     
    std::vector<std::vector<double>> c_xyz;
    for (size_t i = 0; i < natoms; i++) {
        if (atom_type[i] == std::string("C")) 
            c_xyz.push_back(xyz[i]);
    }
    xyz = c_xyz;
}

double find_dist(std::vector<double> &xyz1, std::vector<double> &xyz2) {
    double dist = 0.0;
    for (auto it1 = xyz1.begin(), it2 = xyz2.begin(); (it1 != xyz1.end() || it2 != xyz2.end()) ; it1++, it2++) {
        dist += std::pow(*it1 - *it2, 2);
    }
    return std::sqrt(dist);
}

void find_adjacency(std::vector<std::vector<double>> &xyz, Eigen::MatrixXd &adjacency) { 
    size_t n_cs = xyz.size();
    for (size_t i = 0; i < n_cs; i++) 
        for (size_t j = 0; j < n_cs; j++) {
            adjacency(i,j) = find_dist(xyz[i], xyz[j]);
        }
}

void strip_down(Eigen::MatrixXd &adjacency, double cutoff) {
    assert(adjacency.rows() == adjacency.cols()); // make sure it is square!
    size_t n = adjacency.rows();
    for (size_t i = 0; i < n; i++) 
        for (size_t j = 0; j < n; j++) {
            if (adjacency(i,j) > cutoff) // if far away
                adjacency(i,j) = 0.0; // zero it out
            else if (adjacency(i,j) > 1e-6) // some positive noise
                adjacency(i,j) = 1.0;
        }
}

void output_csv(std::string &outfile, Eigen::MatrixXd &adjacency){
    std::ofstream outstream(outfile.c_str());
    const static Eigen::IOFormat CSVFormat(Eigen::StreamPrecision, Eigen::DontAlignCols, ", ", "\n");
    outstream << adjacency.format(CSVFormat);
}


